File: models/product.py
from app import ma, db

class Products(db.Model):
    id = db.Column(db.Integer, primary_key = True)
    name = db.Column(db.String(50))
    category = db.Column(db.String(100))
    created_at = db.Column(db.DateTime)
    updated_at = db.Column(db.DateTime)
    country = db.Column(db.String(50))
    materials = db.Column(db.String(1000))
    image_url = db.Column(db.String(1000))
    taste = db.Column(db.String(1000))
    aroma = db.Column(db.String(1000))
    weight = db.Column(db.String(10))
    quantity = db.Column(db.String(5))
    price = db.Column(db.String(20))

    def __init__(self, name, category, created_at, updated_at, country, materials, image_url, taste, aroma, weight, quantity, price):
        self.name = name
        self.category = category
        self.created_at = created_at
        self.updated_at = updated_at
        self.country = country
        self.materials = materials
        self.image_url = image_url
        self.taste = taste
        self.aroma = aroma
        self.weight = weight
        self.quantity = quantity
        self.price = price

class ProductSchema(ma.Schema):
    class Meta:
        fields = ("id", "name", "category", "created_at", "updated_at", "country", "materials", "image_url", "taste", "aroma", "weight", "quantity", "price")

product_schema = ProductSchema()
products_schema = ProductSchema(many = True)

File: scripts/typescript/wish-list/product.ts
let productOpen = false;

function openCloseProduct(targetId: string): void {
    let product = document.getElementById(targetId);
    productOpen = !productOpen;
    if (productOpen) {
        product.classList.add("product_bodyOpen");
        product.classList.remove("product_bodyClosed");
    } else {
        product.classList.add("product_bodyClosed");
        product.classList.remove("product_bodyOpen");
    };
};

function moveProductToReviews(id: number): void {
    let name = (document.getElementById("productName" + id) as HTMLInputElement)?.value;
    let category = (document.getElementById("productCategory" + id) as HTMLInputElement)?.value;
    let country = (document.getElementById("productCountry" + id) as HTMLInputElement)?.value;
    let materials = (document.getElementById("productMaterials" + id) as HTMLInputElement)?.value;
    let image_url = (document.getElementById("productImageUrl" + id) as HTMLInputElement)?.value;
    let taste = (document.getElementById("productTaste" + id) as HTMLInputElement)?.value;
    let aroma = (document.getElementById("productAroma" + id) as HTMLInputElement)?.value;
    let weight = (document.getElementById("productWeight" + id) as HTMLInputElement)?.value;
    let quantity = (document.getElementById("productQuantity" + id) as HTMLInputElement)?.value;
    let price = (document.getElementById("productPrice" + id) as HTMLInputElement)?.value;

    fetch(`/api/reviews`, {
        method: "POST",
        headers: {
            "Accept": "application/json",
            "Content-Type": "application/json"
        },
        body: JSON.stringify({
            name: name,
            category: category,
            country: country,
            materials: materials,
            image_url: image_url,
            taste: taste,
            aroma: aroma,
            weight: weight,
            quantity: quantity,
            price: price,
            note: "",
            vote: 0,
            created_by: localStorage.getItem("user_id")
        })
    })
    .then(() => {
        alert("Review added successfully");
    })
    .then(() => deleteProductFromWishList(id))
    .then(() => window.location.href = `/reviews?user=${localStorage.getItem("user_id")}`)
    .catch(err => console.log(err));
};

function updateProductInWishList(id: number): void {
    let name = (document.getElementById("productName" + id) as HTMLInputElement)?.value;
    let category = (document.getElementById("productCategory" + id) as HTMLInputElement)?.value;
    let country = (document.getElementById("productCountry" + id) as HTMLInputElement)?.value;
    let materials = (document.getElementById("productMaterials" + id) as HTMLInputElement)?.value;
    let image_url = (document.getElementById("productImageUrl" + id) as HTMLInputElement)?.value;
    let taste = (document.getElementById("productTaste" + id) as HTMLInputElement)?.value;
    let aroma = (document.getElementById("productAroma" + id) as HTMLInputElement)?.value;
    let weight = (document.getElementById("productWeight" + id) as HTMLInputElement)?.value;
    let quantity = (document.getElementById("productQuantity" + id) as HTMLInputElement)?.value;
    let price = (document.getElementById("productPrice" + id) as HTMLInputElement)?.value;

    fetch(`/api/products/${id}`, {
        method: "PUT",
        headers: {
            "Accept": "application/json",
            "Content-Type": "application/json"
        },
        body: JSON.stringify({
            name: name,
            category: category,
            country: country,
            materials: materials,
            image_url: image_url,
            taste: taste,
            aroma: aroma,
            weight: weight,
            quantity: quantity,
            price: price,
            created_by: localStorage.getItem("user_id")
        })
    })
    .then(() => {
        alert("Product updated successfully");
    })
    .then(() => window.location.reload())
    .catch(err => console.log(err));
};

function deleteProductFromWishList(id: number): void {
    fetch(`/api/products/${id}`, {
        method: "DELETE",
        headers: {
            "Accept": "application/json",
            "Content-Type": "application/json"
        }
    })
    .then(() => {
        alert("Product deleted successfully");
        window.location.reload();
    })
    .catch(err => console.log(err));
};